### PR TITLE
Add Iterator::{sum_nonempty, product_nonempty}

### PR DESCRIPTION
**Pull request by Emerentius**
This implements `{Sum<U>, Product<U>} for Option<T>` such that it returns `None` if the iter is empty.
That allows one to distinguish between an empty iter and iters that sum/multiply to their neutral element.

There's an alternative way of summing into `Option<T>`, which doesn't require a neutral starting element and therefore no `T: Sum / Product` bound. It works by taking the first element and then adding / multiplying the rest onto that, but it can't accomodate the case of different types such as `Iterator<Item = U> -> Option<T>`. This PR keeps in line with the rest of the trait impls and therefore does require `T: Sum<U>`.

This impl is for a stable trait and would therefore be insta-stable if merged.

### CODE REVIEW

**Comment by clarcharr** [comment]
Invoking `once(first).chain(iter)` isn't a very good idea, IMHO. Also, I'm not sure if `Option` is the right type for this.

The way this should be done is through a separate iterator adaptor, like:

```
pub struct IteratedCheck<I> {
    iter: I,
    iterated: bool,
}
impl<I: Iterator> Iterator for I {
    type Item = <I as Iterator>::Item;
    #[inline]
    fn next(&mut self) -> Option<Self::Item> {
        self.iterated = true;
        self.iter.next()
    }
    // forward other methods too
}
impl<T, U> Sum<U> for Option<T>
where
    T: Sum<U>,
{
    #[inline]
    fn sum<I: Iterator<Item = U>>(iter: I) -> Option<T> {
        let mut adapter = IteratedCheck { iter, iterated: false };
        Some(adapter.by_ref().sum()).filter(adapter.iterated)
    }
}
// and product too
```

This allows you to avoid any performance burdens from `chain`.

File: src/libcore/iter/traits.rs [hunk 1]
```diff
@@ -836,6 +837,52 @@ macro_rules! float_sum_product {
 integer_sum_product! { i8 i16 i32 i64 i128 isize u8 u16 u32 u64 u128 usize }
 float_sum_product! { f32 f64 }
 
+#[stable(feature = "iter_arith_traits_option", since="1.29.0")]
```
Comment (kennytm) [inline src/libcore/iter/traits.rs]: Should be 1.28.0 😄

**Comment by kennytm** [comment]
highfive failed to assign anybody again cc @rust-lang/infra. 

r? @sfackler

File: src/libcore/iter/traits.rs (hunk 1 above)
Comment (Emerentius) [reply src/libcore/iter/traits.rs]: Heh, just planning ahead. I'm sure the decision process will delay this at least 4 weeks given that it's insta-stable.

**Comment by pietroalbini** [comment]
Ping from triage @sfackler! This PR needs your review.

**Comment by sfackler** [comment]
This seemed a bit weird to me at first, but I think seems reasonable.

@rfcbot fcp merge

**Comment by clarcharr** [comment]
As I stated in my comment above, this should use an iterator adapter instead of using `chain` to avoid performance issues.

**Comment by sfackler** [comment]
Yeah, the implementation is separate from the impl itself.

**Comment by scottmcm** [comment]
Similarly to the analogy to `Result`, would this conflict with a hypothetical future `impl Add<Option<U>> for Option<T>`?  Like `fold` vs `fold1`, I personally would have expected the behaviour difference here to be denoted by a different function name, not just a different impl.

**Comment by Emerentius** [comment]
(clarification from irc: scottmcm is asking about a potential impl such that one could `sum` or `product` for types implementing `Add` or `Mul`)
I've mentioned this in the OP, it is possible to impl `{Sum, Product}<T> for Option<T> where T: {Add, Mul}`. That implementation doesn't allow different types `U` and `T` because there are no traits other than `{Sum, Product}` that offer us a neutral starting element. An iterator of length 1 has nothing to add | multiply to the first element of type `U` to get `T`.

For the same reason, the lack of a starter element means that an `Option<T>`-like type is the only viable target for `{Sum, Product}<T> where T: {Add, Mul}`

**Comment by scottmcm** [comment]
I want to reiterate what @Emerentius said above:

> For completeness' sake I should mention that there is also an implementation of Sum<Result<U, E>> for Result<T, E> where T: Sum<U> which short-circuits on errors. A corresponding impl that short-circuits on None would conflict with this one.

I think that, with both `Option` and `Result` being usable with `?`, it's surprising to have them behave so differently here.

Concretely, I propose taking the impls from here and putting them in a new `sum1` method on `Iterator` instead.

**Comment by clarcharr** [comment]
I agree with choosing another method on `Iterator`. Bikeshedding on the name can happen after it's implemented, although I'd like to point out that 1-suffixed convention that Haskell uses probably would not be the best way to go about things, as a 1 is not easily noticeable at first glance. `nonempty_sum` might be better.

**Comment by shepmaster** [comment]
> 1-suffixed convention

I'm not in love with it, but Itertools does make use of it:

- [`fold1`](https://docs.rs/itertools/0.7.8/itertools/trait.Itertools.html#method.fold1)
- [`tree_fold1`](https://docs.rs/itertools/0.7.8/itertools/trait.Itertools.html#method.tree_fold1)

**Comment by clarcharr** [comment]
@shepmaster I'd argue that `itertools` is simply following the convention directly from Haskell, so, imho that isn't really a compelling argument other than precedence. Which usually is a compelling argument, but :woman_shrugging:

**Comment by Emerentius** [comment]
I have no opinion on `Sum` impl or new method. Both would be fine.

If we're adding `sum1`, then I would really like to see a way to `sum1` for any `T: Add`. It's the perfect return type for that after all. But it seems that, at the moment, we couldn't do that and also keep it consistent with regular `sum`.

If we defined a new trait `Sum1` and do blanket impls such as `impl<T: Add> Sum1 for T` and `impl<T, U: Sum<T>> Sum1<T> for U`, these two would overlap and be forbidden. Neither is more specific than the other and so they can't be specialized either.

In any case, if `Option` returning special methods are added, the general purpose `fold1` should also be part of it.

**Comment by pietroalbini** [comment]
Ping from triage @sfackler @Emerentius! How should we move forward with this PR?

**Comment by Emerentius** [comment]
Sorry, I forgot about this PR for a while.

Thinking about it again, I don't see why `Sum<T> for Option<T>` and `Sum<Option<T>> for Option<T>` couldn't co-exist. It would simply mean that an `Iterator<Item=Option<T>>` could sum into either `Option<T>` or `Option<Option<T>>`.
Awesome for type inference (not), but otherwise [perfectly legal](http://play.rust-lang.org/?gist=233188ff5da98fc4bc360b29a953d4f6&version=stable&mode=debug&edition=2015).

Summing into `Option<Option<T>>` has the esoteric semantics where the outer `Option` shows nonemptyness of the iter, the inner is the result of a short-circuiting sum. I could see someone confused about their types trial-and-erroring themselves into this by mistake, but I don't think it's a big issue.

I think that's much preferable to having a special method that still couldn't sum `T: Add` because of having to maintain consistency with the `Sum` trait.

I can implement both of the above trait impls.

**Comment by Emerentius** [comment]
hmm.. I thought I just did add that

**Comment by alexcrichton** [comment]
Thanks @Emerentius for the updates! I wonder if we can perhaps think of a different name other than `*_nonempty` though? That name implies to me sort of that it's somehow asserting it's nonempty, in that I'd expect the `sum` and `sum_nonempty` type signatures to be switched. Something like `sum_or_none` seems like a bit wordy though...

**Comment by shepmaster** [comment]
> The *1 suffixing has weak precedence (Haskell) 
> Edit: nom also uses the fold1 terminology

As [does itertools](https://github.com/rust-lang/rust/pull/50884#issuecomment-399559733).

**Comment by alexcrichton** [comment]
To make sure I understand, you're thinking we could add `sum1` as a method which returns `None` if the iterator is empty?

**Comment by Emerentius** [comment]
Either `sum1` or `sum_nonempty` is fine by me. I was trying to evade the bikeshed.

**Comment by alexcrichton** [comment]
Do others from @rust-lang/libs have thoughts on this? I personally sort of feel like this functionality belongs in `itertools` rather than the standard library, but if others feel it should stay we may need some help with naming!

**Comment by scottmcm** [comment]
Hmm, definitely feels hard to argue that core should have `sum1` xor `fold1`.

**Comment by TimNN** [comment]
Ping from triage, @sfackler, @alexcrichton & @Emerentius. What is the status of this PR? It looks like there's been an FCP but there are still some open questions.

**Comment by TimNN** [comment]
Ping from triage again, @sfackler, @alexcrichton & @Emerentius!

**Comment by Emerentius** [comment]
Given that the reception is just lukewarm, it probably is best to put it in itertools for now.
There's also still the downside with the current design that `T: {Add | Mul}` don't work with it. Maybe we'll get the necessary features for overlapping trait impls in the future.

**Comment by alexcrichton** [comment]
That sounds reasonable to me! @Emerentius shall we close this PR in that case?

**Comment by Emerentius** [comment]
Opened https://github.com/bluss/rust-itertools/pull/300.
